tests/rest/client/test_login.py

```diff
@@ -384,6 +384,37 @@ def test_session_can_hard_logout_all_sessions_after_being_soft_logged_out(
         channel = self.make_request(b"POST", "/logout/all", access_token=access_token)
         self.assertEqual(channel.result["code"], b"200", channel.result)
 
+    def test_login_with_overly_long_device_id_fails(self) -> None:
+        self.register_user("mickey", "cheese")
+
+        # create a device_id longer than 512 characters
+        device_id = "yolo" * 512
+
+        body = {
+            "type": "m.login.password",
+            "user": "mickey",
+            "password": "cheese",
+            "device_id": device_id,
+        }
+
+        # make a login request with the bad device_id
+        channel = self.make_request(
+            "POST",
+            "/_matrix/client/r0/login",
+            json.dumps(body).encode("utf8"),
+            custom_headers=None,
+        )
+
+        # test that the login fails with the correct error code/message
+        self.assertEqual(channel.code, 400)
+        self.assertEqual(
+            channel.json_body,
+            {
+                "errcode": "M_INVALID_PARAM",
+                "error": "Device_id cannot be longer than 512 characters.",
```

Comment:
I tend to be against checking localised error text in tests, as:

1. If we update the text, we'll need to do so in multiple places
2. Clients should not be performing logic based on the text anyways, as it may change.

As such, I think we should only check the `errcode` is correct here.

Reply:
Right, thanks for explaining that. Makes sense!